bundles/action/org.openhab.action.pushover/src/main/java/org/openhab/action/pushover/internal/Pushover.java

```diff
@@ -312,33 +339,64 @@ public static boolean pushover(@ParamDoc(name = "apiKey", text = "Your applicati
             }
 
             if (!StringUtils.isEmpty(sound)) {
-                addEncodedParameter(data, MESSAGE_KEY_SOUND, sound);
+                parts.add(new StringPart(MESSAGE_KEY_SOUND, sound));
             } else if (!StringUtils.isEmpty(defaultSound)) {
-                addEncodedParameter(data, MESSAGE_KEY_SOUND, defaultSound);
+                parts.add(new StringPart(MESSAGE_KEY_SOUND, defaultSound));
             }
 
             if (isValueInList(API_HIGH_PRIORITY_LIST, priority)) {
                 if (retry >= API_MIN_RETRY_SECONDS) {
-                    addEncodedParameter(data, MESSAGE_KEY_RETRY, String.valueOf(retry));
+                    parts.add(new StringPart(MESSAGE_KEY_RETRY, String.valueOf(retry)));
                 } else {
                     logger.warn("Retry value of {} is too small. Using default value of {}.", retry,
                             API_MIN_RETRY_SECONDS);
-                    addEncodedParameter(data, MESSAGE_KEY_RETRY, String.valueOf(API_MIN_RETRY_SECONDS));
+                    parts.add(new StringPart(MESSAGE_KEY_RETRY, String.valueOf(API_MIN_RETRY_SECONDS)));
                 }
 
                 if (expire <= API_MAX_EXPIRE_SECONDS) {
-                    addEncodedParameter(data, MESSAGE_KEY_EXPIRE, String.valueOf(expire));
+                    parts.add(new StringPart(MESSAGE_KEY_EXPIRE, String.valueOf(expire)));
                 } else {
                     logger.warn("Expire value of {} is too large. Using default value of {}.", expire,
                             API_MAX_EXPIRE_SECONDS);
-                    addEncodedParameter(data, MESSAGE_KEY_EXPIRE, String.valueOf(API_MAX_EXPIRE_SECONDS));
+                    parts.add(new StringPart(MESSAGE_KEY_EXPIRE, String.valueOf(API_MAX_EXPIRE_SECONDS)));
                 }
             }
+            if (attachment != null) {
+                logger.info("Push the image attachment <{}>.", attachment);
+                try {
+                    File f = new File(attachment);
+                    FilePart fp = new FilePart(MESSAGE_KEY_ATTACHMENT, f.getName(), f);
+                    fp.setContentType("image/jpeg");
+                    parts.add(fp);
+                } catch (Throwable t) {
```

Comment:
for what reason?
i guess, we can have a lot of runtime exceptions here...
i can reduce the Throwable to RuntimeException

Reply:
What about adding `IOException`? the `FileNotFoundException` is an instance of `IOException` and not an instance of `RuntimeException`. The bundle should handle those correctly.
```
2018-03-29 12:04:28.615 [ERROR] [ab.action.pushover.internal.Pushover] - An error occurred while notifying your mobile device.
java.io.FileNotFoundException: File is not readable.
	at org.apache.commons.httpclient.methods.multipart.FilePartSource.<init>(FilePartSource.java:72) ~[?:?]
	at org.apache.commons.httpclient.methods.multipart.FilePartSource.<init>(FilePartSource.java:89) ~[?:?]
	at org.apache.commons.httpclient.methods.multipart.FilePart.<init>(FilePart.java:160) ~[?:?]
	at org.openhab.action.pushover.internal.Pushover.pushover(Pushover.java:368) ~[?:?]
	at org.openhab.action.pushover.internal.Pushover.sendPushoverImage(Pushover.java:107) ~[?:?]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:?]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:?]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:?]
	at org.eclipse.xtext.xbase.interpreter.impl.XbaseInterpreter.invokeOperation(XbaseInterpreter.java:1085) ~[?:?]
	at org.eclipse.xtext.xbase.interpreter.impl.XbaseInterpreter.invokeOperation(XbaseInterpreter.java:1060) ~[?:?]
	at org.eclipse.xtext.xbase.interpreter.impl.XbaseInterpreter._invokeFeature(XbaseInterpreter.java:1046) ~[?:?]
	at org.eclipse.xtext.xbase.interpreter.impl.XbaseInterpreter.invokeFeature(XbaseInterpreter.java:991) ~[?:?]
	at org.eclipse.smarthome.model.script.interpreter.ScriptInterpreter.invokeFeature(ScriptInterpreter.java:143) ~[?:?]
	at org.eclipse.xtext.xbase.interpreter.impl.XbaseInterpreter._doEvaluate(XbaseInterpreter.java:901) ~[?:?]
	at org.eclipse.xtext.xbase.interpreter.impl.XbaseInterpreter._doEvaluate(XbaseInterpreter.java:864) ~[?:?]
	at org.eclipse.xtext.xbase.interpreter.impl.XbaseInterpreter.doEvaluate(XbaseInterpreter.java:223) ~[?:?]
	at org.eclipse.smarthome.model.script.interpreter.ScriptInterpreter.doEvaluate(ScriptInterpreter.java:219) ~[?:?]
	at org.eclipse.xtext.xbase.interpreter.impl.XbaseInterpreter.internalEvaluate(XbaseInterpreter.java:203) ~[?:?]
	at org.eclipse.xtext.xbase.interpreter.impl.XbaseInterpreter._doEvaluate(XbaseInterpreter.java:446) ~[?:?]
	at org.eclipse.xtext.xbase.interpreter.impl.XbaseInterpreter.doEvaluate(XbaseInterpreter.java:227) ~[?:?]
	at org.eclipse.smarthome.model.script.interpreter.ScriptInterpreter.doEvaluate(ScriptInterpreter.java:219) ~[?:?]
	at org.eclipse.xtext.xbase.interpreter.impl.XbaseInterpreter.internalEvaluate(XbaseInterpreter.java:203) ~[?:?]
	at org.eclipse.xtext.xbase.interpreter.impl.XbaseInterpreter.evaluate(XbaseInterpreter.java:189) ~[?:?]
	at org.eclipse.smarthome.model.script.runtime.internal.engine.ScriptImpl.execute(ScriptImpl.java:82) ~[?:?]
	at org.eclipse.smarthome.model.rule.runtime.internal.engine.RuleEngineImpl.lambda$2(RuleEngineImpl.java:339) ~[?:?]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [?:?]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:?]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180) [?:?]
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293) [?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:?]
	at java.lang.Thread.run(Thread.java:748) [?:?]
```

Reply:
whats wrong with the curernt exception handling?
the only thing, one could change here, is to avoid the stack trace, a matter of taste
what do you think, what would be the difference in the handling of those different exceptions
IOException, FileNotFoundException, FooBarException ?

Reply:
A catch statement should never catch throwable since it can hide unpredictable errors.

I don't know if we already have listed it in the coding guidelines but our Static Code Analysis Tools gives a warning in that case.

You can easily change it to `catch (IOException | RuntimeException t)`.

Reply:
changed

Reply:
Code should generally not catch `RuntimeException`s.

I would ask this be removed.

Reply:
no RuntimeException's any more